pkg/store/s3/s3.go

```diff
@@ -0,0 +1,146 @@
+	ListModuleVersions(ctx context.Context, namespace, name, system string) ([]*core.ModuleVersion, error)
+	GetModuleVersion(ctx context.Context, namespace, name, system, version string) (*core.ModuleVersion, error)
+}
+
+// S3Store implements S3StoreInterface
+type S3Store struct {
+	client s3iface.S3API
+	cache  map[string][]*core.ModuleVersion
+	region string
+	bucket string
+	logger *zap.Logger
+	mut    sync.RWMutex
+}
+
+func NewS3Store(client s3iface.S3API, region string, bucket string, logger *zap.Logger) *S3Store {
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+
+	return &S3Store{
+		client: client,
+		cache:  make(map[string][]*core.ModuleVersion),
+		region: region,
+		bucket: bucket,
+		logger: logger,
+	}
+}
+
+func (s *S3Store) ListModuleVersions(ctx context.Context, namespace, name, system string) ([]*core.ModuleVersion, error) {
+	s.mut.RLock()
+	defer s.mut.RUnlock()
+
+	addr := filepath.Join(namespace, name, system)
+	vers, err := s.fetchModuleVersions(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return vers, nil
+}
+
+func (s *S3Store) fetchModuleVersions(ctx context.Context, address string) ([]*core.ModuleVersion, error) {
+	p := address + "/"
+	in := &s3.ListObjectsV2Input{
+		Bucket: aws.String(s.bucket),
+		Prefix: aws.String(p),
+	}
+	out, err := s.client.ListObjectsV2WithContext(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+
+	vers := make([]*core.ModuleVersion, 0)
+	for _, o := range out.Contents {
+		path := o.Key
+		if isValidModuleSourcePath(*path) {
+			vers = append(vers, &core.ModuleVersion{
+				Version:   strings.Split(*path, "/")[3],
+				SourceURL: fmt.Sprintf("s3::https://%s.s3.%s.amazonaws.com/%s", s.bucket, s.region, *path),
+			})
+		}
+	}
+
+	s.cache[address] = vers
+
+	return vers, nil
+}
+
+func (s *S3Store) GetModuleVersion(ctx context.Context, namespace, name, system, version string) (*core.ModuleVersion, error) {
+	s.mut.RLock()
+	defer s.mut.RUnlock()
+
+	addr := filepath.Join(namespace, name, system)
+	ver, err := s.fetchModuleVersion(ctx, addr, version)
+	if err != nil {
+		return nil, err
+	}
+
+	return ver, nil
+}
+
+func (s *S3Store) fetchModuleVersion(ctx context.Context, address, version string) (*core.ModuleVersion, error) {
+	vers := s.cache[address]
+	for _, o := range vers {
+		if o.Version == version {
+			return o, nil
+		}
+	}
+
+	path := path.Join(address, version)
+	keySuffix := version + ".zip"
+	if !isValidModuleSourcePath(path) {
+		s.logger.Warn("invalid module path requested: " + path)
+		return nil, fmt.Errorf("module version path '%s' is not valid", path)
+	}
+	_, err := s.client.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path + "/" + keySuffix),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ver := &core.ModuleVersion{
+		Version:   version,
+		SourceURL: fmt.Sprintf("s3::https://%s.s3.%s.amazonaws.com/%s/%s", s.bucket, s.region, path, keySuffix),
+	}
+
+	s.cache[address] = append(vers, ver)
```

Comment:
This write to cache is not synchronised and might result in multiple identical versions for the same cache key.

Reply:
This looks good now!